### PR TITLE
add some C functions

### PR DESCRIPTION
**Pull request by KouweiLee**
I add some C functions for rukos including:
1. qsort
2. fnmatch
3. remove, mkdir, rmdir, unlink
4. fdopen
5. freopen
6. sendmsg
7. pthread_cond_init, wait, signal, broadcast
8. pthread_mutex_trylock

And I add some tests for them including:
1. filetest: test fdopen ,freopen, remove
2. httpclient: test sendmsg
3. pthread/basic: test pthread_cond_xxx

And I fix one bug for envtest so that it will exit when error occurs.

### CODE REVIEW

File: modules/axhal/src/platform/x86_pc/mod.rs
Comment (KouweiLee) [inline modules/axhal/src/platform/x86_pc/mod.rs:1]: To make `rust_entry` clearer, I take out the cmdline parsing part as the function `parse_cmdline`

File: ulib/axlibc/c/fnmatch.c [hunk 1]
```diff
@@ -9,16 +9,268 @@
...
 #define QUESTION    -4
 #define STAR        -5
 
-// TODO
+static int str_next(const char *str, size_t n, size_t *step)
+{
+	if (!n) {
+		*step = 0;
+		return 0;
+	}
```
Comment (coolyjg) [inline ulib/axlibc/c/fnmatch.c:26]: In musl libc, there are more code like `str[0] >= 128U`. Why ignore these codes?  Or add comment `TODO` here.

File: ulib/axlibc/c/fnmatch.c [hunk 2]
```diff
@@ -9,16 +9,268 @@
...
+static int casefold(int k)
+{
+    int c = toupper(k);
+    return c == k ? tolower(k) : c;
+    
+}
+
+static int match_bracket(const char *p, int k, int kfold)
+{
+	// wchar_t wc;
```
Comment (coolyjg) [inline ulib/axlibc/c/fnmatch.c:80]: Why are there many comment lines? Add comment or `TODO` here.

File: ulib/axlibc/c/socket.c [hunk 3]
```diff
@@ -46,11 +47,47 @@ int setsockopt(int fd, int level, int optname, const void *optval, socklen_t opt
     return 0;
 }
 
-// TODO
+// TODO: remove this function in future work
+ssize_t ax_sendmsg(int fd, const struct msghdr *msg, int flags);
+
+#define __CMSG_LEN(cmsg) (((cmsg)->cmsg_len + sizeof(long) - 1) & ~(long)(sizeof(long) - 1))
```
Comment (coolyjg) [inline ulib/axlibc/c/socket.c:53]: These codes should be put into sys/socket.h. Please follow musl libc.

File: ulib/axlibc/include/ax_pthread_cond.h
Comment (coolyjg) [inline ulib/axlibc/include/ax_pthread_cond.h:1]: Add this file to gitignore

File: api/arceos_posix_api/src/imp/pthread/condvar.rs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+/// Initialize a condition variable
+pub unsafe fn sys_pthread_cond_init(
+    condvar: *mut ctypes::pthread_cond_t,
+    _attr: *mut ctypes::pthread_condattr_t,
+) -> c_int {
+    debug!("sys_pthread_cond_init <= {:#x}", condvar as usize);
+    syscall_body!(sys_pthread_cond_init, {
+        // TODO: Do we need a ManualDrop???
```
Comment (coolyjg) [inline api/arceos_posix_api/src/imp/pthread/condvar.rs:61]: No, I think follow pthread_mutex_init is ok

File: apps/c/filetest/main.c [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        return -1;
+	}
+    ret = remove("filetest");
+	if (ret == -1){
+		perror("remove dir error");
+        return -1;
+	}
+	puts("filetest success!");
+	return 0;
+}
```
Comment (coolyjg) [inline apps/c/filetest/main.c:49]: reserve a blank line in the end of this file

**Review by coolyjg — COMMENTED** [review]
Good job! But I need some time to do more checks. Please fix a conflict.

**Review by coolyjg — COMMENTED** [review]
Good job! But I need some time to do more checks. Please fix a conflict.

File: ulib/axlibc/c/fnmatch.c (hunk 1 above)
Comment (KouweiLee) [reply ulib/axlibc/c/fnmatch.c:26]: Now that Rukos has not implemented wide characters, these codes can be added after Rukos implements wide characters

File: apps/c/filetest/main.c (hunk 5 above)
Comment (KouweiLee) [reply apps/c/filetest/main.c:49]: ok, but I want to know why we should reserve a blan line?

**Review by coolyjg — APPROVED** [review]

**Pull request merged**